libs/dotnet-sdk/FeatureBoardLastCheckedService.cs

```diff
@@ -0,0 +1,47 @@
+using FeatureBoard.DotnetSdk.Models;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+
+namespace FeatureBoard.DotnetSdk;
+
+public delegate ref DateTimeOffset LastCheckedTimeProvider();
+
+internal class FeatureBoardLastCheckedService : IFeatureBoardService
+{
+  private readonly FeatureBoardService _innerFeatureBoardService;
+  private readonly IOptions<FeatureBoardOptions> _options;
+  private readonly DateTimeOffset _now;
+  private readonly ILogger _logger;
+  private readonly LastCheckedTimeProvider _lastCheckedTimeProvider;
+
+  public FeatureBoardLastCheckedService(FeatureBoardService innerFeatureBoardService, IOptions<FeatureBoardOptions> options, Func<DateTimeOffset> now, LastCheckedTimeProvider lastCheckedTimeProvider, ILogger<FeatureBoardLastCheckedService> logger)
+  {
+    _innerFeatureBoardService = innerFeatureBoardService;
+    _options = options;
+    _lastCheckedTimeProvider = lastCheckedTimeProvider;
+    _now = now();
+    _logger = logger;
+  }
+
+  public virtual async Task<bool?> RefreshFeatureConfiguration(CancellationToken cancellationToken)
+  {
+    var maxAgeHasExpired = _now > _lastCheckedTimeProvider() + _options.Value.MaxAge;
+    if (!maxAgeHasExpired)
+    {
+      _logger.LogDebug("Feature Configuration has not reached {0}, skipping refresh", _options.Value.MaxAge);
+      return false;
+    }
+
+    void UpdateLastChecked() // Sync method to allow use of lastChecked ref-local variable
```

Comment:
`ref` is basically C#'s way to allow pass around a not null pointer. Normal assignment will create a new independent variable - this is what happens on line 28 for example. However having an independent variable is of no use when updating a long lived shared one. If the `ref` assignment wasn't there, the change in value on line 38 would get lost as soon as it goes out of scope as it only modifies a clone. Using `ref` means the change in value gets applied to the long lived static variable `RegisterFeatureBoard::lastChecked` (or whatever reference is injected by the provider) instead a copy of it.

Alternative to doing all this `ref` stuff was to create some arbitrary object for the sole purpose of being the container for a `lastCheckedTime`  _property_  ie. something like:
```
class FeatureBoardLastCheckedState 
{
   DateTimeOffset lastChecked
}
```

Happy to amend to this more familiar approach as this `ref` stuff is quite _obtuse_.

The inline "local function" it is not static - can see it is referencing `_lastCheckedTimeProvider` member variable. You can add `static` modifier to local functions though and then it can't reference anything except things passed as a parameter (will get compile error if try to reference anything else).

Will have to think more on the question of possible race condition.

The return value of `_innerFeatureBoardService.RefreshFeatureConfiguration()` will be `null` if the calling thread was unable to enter the semaphore, hence the `null` check on line 42 guarding the `lastChecked` reference being from updated in that situation. However there might be other scenarios I've missed...